### PR TITLE
BUG: DataFrame.pivot drops column level names when both rows and columns are multiindexed

### LINKED ISSUE #36360: BUG: MutliIndex names lost during pivot
- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
d = pd.DataFrame([[1,2,3],[2,2,3],[3,2,3],[4,2,3],[5,2,3]], 
                 columns=pd.MultiIndex.from_tuples([[0, 0, 0], [0, 0, 1], [0, 0, 2]], names=['c1', 'c2', 'c3']),
                 index=pd.MultiIndex.from_tuples([[0, 0, 0, 0, 0, 0, 0],
                                                  [0, 0, 1, 0, 0, 0, 1],
                                                  [0, 1, 0, 0, 0, 1, 0],
                                                  [0, 1, 1, 0, 0, 1, 1],
                                                  [1, 0, 0, 0, 1, 0, 0]
                                                  ], 
                                                 names=['i1', 'i2', 'i3', 'i4', 'i5', 'i6', 'i7']))
e = d.pivot_table(index=['i1',], columns=['i2', 'i3',], values=[(0, 0, 1), (0, 0, 2)], aggfunc=lambda v: v.values[0])
e.columns
```
result is:
```MultiIndex([(0, 0, 1, 0, 0),
            (0, 0, 1, 0, 1),
            (0, 0, 1, 1, 0),
            (0, 0, 1, 1, 1),
            (0, 0, 2, 0, 0),
            (0, 0, 2, 0, 1),
            (0, 0, 2, 1, 0),
            (0, 0, 2, 1, 1)],
           names=[None, None, None, 'i2', 'i3'])
```
some names is None, but must be
```names=['c1', 'c2', 'c3', 'i2', 'i3'])```

#### Problem description
Problems seems in  pandas/core/groupby/generic.py in class DataFrameGroupBy in function  _wrap_aggregated_output
Line
```python
name = self._obj_with_exclusions._get_axis(1 - self.axis).name
```
seems to must be
```python
name_axis = self._obj_with_exclusions._get_axis(1 - self.axis)
name = name_axis.names if isinstance(name_axis, MultiIndex) else name_axis.name
```

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #36360
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Fixed by checking for multi-indexed columns in `DataFrameGroupBy._wrap_aggregated_output`

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1643,8 +1643,12 @@ def _wrap_aggregated_output(
         DataFrame
         """
         indexed_output = {key.position: val for key, val in output.items()}
-        name = self._obj_with_exclusions._get_axis(1 - self.axis).name
-        columns = Index([key.label for key in output], name=name)
+
+        agg_axis = self._obj_with_exclusions._get_axis(1 - self.axis)
+        if isinstance(agg_axis, MultiIndex):
+            columns = Index([key.label for key in output], names=agg_axis.names)
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1649]: does this line alone work? I thought we fixed construction for this. alt does `ensure_index` work here? I *really* do not like special casing like this *any*where except in the index module itself (and even then its not great).

File: pandas/tests/reshape/test_pivot_multilevel.py [hunk 2]
```diff
@@ -190,3 +190,35 @@ def test_pivot_list_like_columns(
         expected_values, columns=expected_columns, index=expected_index
     )
     tm.assert_frame_equal(result, expected)
+
+
+def test_pivot_multiindexed_rows_and_cols():
+    # GH 36360
```
Comment (jreback) [inline pandas/tests/reshape/test_pivot_multilevel.py:196]: do we have appropraite tests if index OR columns are a MI and the other is not?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (arw2019) [reply pandas/core/groupby/generic.py]: It does! I've removed the if-else clause

File: pandas/tests/reshape/test_pivot_multilevel.py (hunk 2 above)
Comment (arw2019) [reply pandas/tests/reshape/test_pivot_multilevel.py:196]: We do (in this file and in `test_pivot.py`)

What existing tests don't cover is the case when `pivot_table` produces a result with multiple MutliIndexed columns - added here

**Comment by arw2019** [comment]
@jreback not sure you'll accept 91f27e1 (but atm it's needed to pass CI)

in 91f27e1 we're passing `Index.names` as the `name` kwarg 

This means that for a simple index we end up with `name=[['index_name']]` in the `Index` constructor so I check for that and set `name=['index_name']`

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -404,6 +404,11 @@ def __new__(
                     )
             # other iterable of some kind
             subarr = com.asarray_tuplesafe(data, dtype=object)
+
+            if isinstance(name, list) and len(name) == 1:
```
Comment (jreback) [inline pandas/core/indexes/base.py:408]: we do not want to add this here, this is a very complicated routine; what are you trying to accomplish?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (arw2019) [reply pandas/core/indexes/base.py]: I was looking for the functionality of `Index._get_names()` / `MultiIndex._get_names()`

Reverted this + using `_get_names` now

**Comment by arw2019** [comment]
@jreback this one is ready for re-review. I now handle simple index and multiindex together with `_set_names(aggregation_axis.names)`

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @arw2019

**Pull request merged**